fix: construct child proxy objects using correct paths and introspection

whitslack opened the pull request:
This PR corrects one bug and one deficiency in `BaseProxyObject.get_children()`:

 * Child proxy objects are constructed using the _parent_ object's path. Thus, all calls to methods or accesses to properties of the child object are erroneously directed to the parent object. So, the first fix in this PR is to construct child proxy objects using the correct paths, formed by concatenating the name of the child to the path of the parent.

 * Child proxy objects are constructed using the child `Node` from the parent's introspection. However, the parent might not provide introspection of its children. The [D-Bus specification](https://dbus.freedesktop.org/doc/dbus-specification.html) says:

   > If a child `<node>` has any sub-elements, then they must represent a complete introspection of the child. If a child `<node>` is empty, then it may or may not have sub-elements; the child must be introspected in order to find out. The intent is that if an object knows that its children are "fast" to introspect it can go ahead and return their information, but otherwise it can omit it.

   When the parent object omits introspection of a child object, the corresponding proxy object is erroneously constructed without any interfaces or children of its own. So, the second fix in this PR is to lazily introspect child objects on first use if the parent's introspection had opted not to provide introspection of the child.

The fixes are validated by two new unit tests in `tests/test_introspection.py`:

 * `test_inline_child()` tests the case that a parent object's introspection _includes_ introspection of its child.
 * `test_noninline_child()` tests the case that a parent object's introspection _omits_ introspection of its child.

whitslack committed:
aio.ProxyObject: fix copy/paste oversight in doc

whitslack committed:
add unit tests of BaseProxyObject.get_children()

These tests currently fail due to one bug and one deficiency in
BaseProxyObject.get_children():

 * Child proxy objects are constructed using the parent object's path.
   Thus, all calls to methods or accesses to properties of the child
   object are erroneously directed to the parent object.

 * Child proxy objects are constructed using the child Node from the
   parent's introspection. However, the parent might not provide
   introspection of its children. The D-Bus specification says:

     “If a child <node> has any sub-elements, then they must represent a
      complete introspection of the child. If a child <node> is empty,
      then it may or may not have sub-elements; the child must be
      introspected in order to find out. The intent is that if an object
      knows that its children are "fast" to introspect it can go ahead
      and return their information, but otherwise it can omit it.”

   When the parent object omits introspection of a child object, the
   corresponding proxy object is erroneously constructed without any
   interfaces or children of its own.

whitslack committed:
BaseProxyObject: construct child proxy objects with correct paths

BaseProxyObject.get_children() was constructing child proxy objects
using the parent object's path. This commit corrects that.

whitslack committed:
BaseProxyObject: lazily introspect children as needed

The D-Bus specification says:

  “If a child <node> has any sub-elements, then they must represent a
   complete introspection of the child. If a child <node> is empty, then
   it may or may not have sub-elements; the child must be introspected
   in order to find out. The intent is that if an object knows that its
   children are "fast" to introspect it can go ahead and return their
   information, but otherwise it can omit it.”

However, BaseProxyObject.get_children() has been assuming that all child
<node> elements provide a complete introspection of their corresponding
child objects. This causes child objects to appear to have no interfaces
or children of their own, even when they in fact do.

To implement the spec, replace BaseProxyObject.introspection with a
lazily-computed property whose value is initialized as the intr.Node
passed to the BaseProxyObject constructor (or parsed by it), if one was
specified, or is otherwise computed upon first use by introspecting the
child object (synchronously). The synchronous introspection required
pulling the introspect_sync method up from glib.MessageBus into
BaseMessageBus as an abstract method and then implementing it in
aio.MessageBus by using asyncio.loop.run_until_complete(…).

pre-commit-ci[bot] committed:
chore(pre-commit.ci): auto fixes

whitslack commented on tests/test_introspection.py:
```diff
@@ -145,3 +150,151 @@ def test_default_interfaces():
+        self.nodes = nodes
+        self.introspect_count = 0
+
+        # disable bus name tracking for testing purposes
+        class MockNameOwners(dict):
+            def get(self, key, default):
+                return dict.get(self, key, default if default else ":")
+
+        self._name_owners = MockNameOwners()
```
I'm sorry, I just don't know how to appease Cython here. I'm subclassing `dict` so I can defeat the name tracking in `BaseProxyObject.get_interface(…)`, but Cython says it's expecting a `dict`, not a `MockNameOwners`. This is beyond my expertise.

bdraco replied:
Cython is going to strongly enforce types. Can you mock it in a different way?

bdraco committed:
chore: small fixes for python 3.9

bdraco committed:
chore: fix introspection test on 3.9

whitslack committed:
chore: alternative name owner tracking defeat to appease Cython

Cython demands that a dict-typed object really be a dict and not a
subclass, presumably because it wants to shortcut the virtual dispatch
and implement all the methods in native code. Thus, my first attempt at
defeating the name owner tracking in MessageBus was a non-starter. This
commit switches to a different strategy: pre-filling the mapping so that
get_interface() won't try to shoot off a GetNameOwner call to the bus.

dlech commented on src/dbus_fast/aio/message_bus.py:
```diff
@@ -318,6 +318,13 @@ async def introspect(
         finally:
             timer_handle.cancel()
 
+    def introspect_sync(
```
It looks like this is being added only for test purposes. It would be better to just access `_loop` in the test (if it is actually needed) instead of adding an API that isn't meant to be used. This is what codecov is complaining that isn't covered because it is mocked in the test anyway. So there doesn't seem to be any reason to add this at all.

dlech replied:
Bah, my browser was being weird and didn't actually find all instance of `introspect_sync` when I was searching for it the first time, so ignore that comment.

dlech commented on src/dbus_fast/aio/message_bus.py:
```diff
@@ -318,6 +318,13 @@ async def introspect(
+    def introspect_sync(
+        self, bus_name: str, path: str, *args: Any, **kwargs: Any
+    ) -> intr.Node:
+        return self._loop.run_until_complete(
```
This is also the line codecov is complaining that is not being covered in any tests.

whitslack replied:
Could the asynchronous `dbus_fast.aio.MessageBus.introspect` method be executed in a thread pool executor via [`loop.run_in_executor`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor)? Of course, we would still need a way to put the calling thread to sleep until the `asyncio.Future` has been completed by the worker thread. Presumably a [done callback](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.add_done_callback) could be used to signal a condition variable (or semaphor, latch, barrier, etc.), but, of course, then the event loop will not be processing any other events while the calling thread is waiting on that synchronization object.

dlech replied:
Actually, I guess the `child_paths` property is new, so just don't add it at all. Just make a method instead.

whitslack replied:
> Actually, I guess the `child_paths` property is new, so just don't add it at all. Just make a method instead.

No, unfortunately it's not new. It was previously a simple data field that the constructor initialized. In order to preserve the existing API but make it lazily-initialized, I had to turn it into a `@property` method. If it's just an ordinary method, then client code that accesses `proxy_obj.child_paths` will suddenly see a bound method instead of a list.

dlech replied:
Maybe AI tools could help write the tests? 

I was also looking at `tests/client/test_aio.py` and it creates a proxy object like this:

```python
    introspection = await bus2.introspect(bus_name, "/test/path")
    bus2.get_proxy_object(bus_name, "/test/path", introspection)
```

This makes me wonder if instead of modifying `BaseProxyObject`, could we solve the issue by adding an optional parameter to `bus2.introspect()` to recursively introspect and return a complete introspection instead? If we want a possibility between no descendant introspection and all descendant introspection, maybe the parameter could be a recursion depth or some sort of pattern matching for child paths (I'm haven't looked deep enough to see what would actually make sense).

whitslack replied:
> I would have the `child_paths` property raise an exception if needs to be introspected and hasn't been yet and create a new async `get_child_paths()` method that can be used instead in cases where this extra introspection is needed.

That will break existing client code.

dlech replied:
Isn't existing client code that uses `get_children()` pretty much broken anyway due to the bugs you mentioned?

whitslack replied:
> This makes me wonder if instead of modifying `BaseProxyObject`, could we solve the issue by adding an optional parameter to `bus2.introspect()` to recursively introspect and return a complete introspection instead?

I suppose that's an acceptable punt, although it could be very expensive in cases where not all of the children will be accessed.

Another idea: keep the existing behavior, where child objects initially appear to have no interfaces and no children of their own, but simply document that the caller will need to overwrite the `introspection` property with the actual introspection of the child object (which they can either have on hand statically as recommended or fetch dynamically via the bus). Writing to the `introspection` property would flush the cached value of `child_paths`, prompting it to be recomputed upon next access.

whitslack replied:
Implemented in 6ed96beb78d4c89df9e8de21f0435f3dd7e986dc.

whitslack committed:
BaseProxyObject: do not lazily introspect children

Make the user do it if they need child objects to be populated.

See: https://github.com/Bluetooth-Devices/dbus-fast/pull/495#pullrequestreview-3194620590

pre-commit-ci[bot] committed:
chore(pre-commit.ci): auto fixes

bdraco committed:
Merge branch 'main' into fix-get_children

dlech reviewed (commented):
Coincidentally, I was writing a new D-Bus service and client the other day, so it had me thinking about this. And now I have a little better understanding of the mechanics.

Based on that new understanding, I would tweak this a bit. Instead of adding an `introspection` property a and making it writable, I would make `get_children()` raise an error if the children only have paths and no interfaces.

Then I would add a new `child_paths` property that returns just the child paths that can be used like:

```python
    for child_path in parent_proxy.child_paths:
        child_node = await bus.introspect(
            "org.example.name", child_path
        )

        child_obj = bus.get_proxy_object(
            "org.example.name", child_path, child_node
        )

        ...
```

This avoids the awkwardness of having `get_children()` return objects that are not fully initialized. And it gives a bit more flexibility in that the user doesn't have to create a proxy for all children if they only need certain ones.

We could also add a new:

```python
async def introspect_and_get_children() -> list[ProxyObject]:
    """
    Introspect each child path and return proxy objects for all child nodes.

    Use this method in cases where the parent object's introspection data
    doesn't contain the full child introspection data. Otherwise you
    can use :meth:`get_children()` which avoids additional bus calls to get
    additional introspection data.
    """
```

To make it convenient for users who really do want all child objects.

bdraco committed:
Merge remote-tracking branch 'origin/main' into fix-get_children

# Conflicts:
#	tests/test_introspection.py
